resolvers/deno/factory.rs

```diff
@@ -0,0 +1,893 @@
+  pub node_modules_dir: Option<NodeModulesDirMode>,
+  pub no_npm: bool,
+  /// The process sate if using ext/node and the current process was "forked".
+  /// This value is found at `deno_lib::args::NPM_PROCESS_STATE`
+  /// but in most scenarios this can probably just be `None`.
+  pub npm_process_state: Option<NpmProcessStateOptions>,
+  pub vendor: Option<bool>,
+}
+
+#[allow(clippy::disallowed_types)]
+pub type WorkspaceFactoryRc<TSys> =
+  crate::sync::MaybeArc<WorkspaceFactory<TSys>>;
+
+pub struct WorkspaceFactory<
+  TSys: EnvCacheDir
+    + EnvHomeDir
+    + EnvVar
+    + EnvCurrentDir
+    + FsCanonicalize
+    + FsCreateDirAll
+    + FsMetadata
+    + FsOpen
+    + FsRead
+    + FsReadDir
+    + FsRemoveFile
+    + FsRename
+    + SystemRandom
+    + SystemTimeNow
+    + ThreadSleep
+    + std::fmt::Debug
+    + MaybeSend
+    + MaybeSync
+    + Clone
+    + 'static,
+> {
+  sys: TSys,
+  deno_dir_path: DenoDirPathProviderRc<TSys>,
+  global_http_cache: Deferred<GlobalHttpCacheRc<TSys>>,
+  http_cache: Deferred<HttpCacheRc>,
+  node_modules_dir_path: Deferred<Option<PathBuf>>,
+  npm_cache_dir: Deferred<NpmCacheDirRc>,
+  npmrc: Deferred<ResolvedNpmRcRc>,
+  node_modules_dir_mode: Deferred<NodeModulesDirMode>,
+  workspace_directory: Deferred<WorkspaceDirectoryRc>,
+  initial_cwd: PathBuf,
+  options: WorkspaceFactoryOptions<TSys>,
+}
+
+impl<
+    TSys: EnvCacheDir
+      + EnvHomeDir
+      + EnvVar
+      + EnvCurrentDir
+      + FsCanonicalize
+      + FsCreateDirAll
+      + FsMetadata
+      + FsOpen
+      + FsRead
+      + FsReadDir
+      + FsRemoveFile
+      + FsRename
+      + SystemRandom
+      + SystemTimeNow
+      + ThreadSleep
+      + std::fmt::Debug
+      + MaybeSend
+      + MaybeSync
+      + Clone
+      + 'static,
+  > WorkspaceFactory<TSys>
+{
+  pub fn new(
+    sys: TSys,
+    initial_cwd: PathBuf,
+    mut options: WorkspaceFactoryOptions<TSys>,
+  ) -> Self {
+    Self {
+      deno_dir_path: options.deno_dir_path_provider.take().unwrap_or_else(
+        || {
+          new_rc(DenoDirPathProvider::new(
+            sys.clone(),
+            DenoDirPathProviderOptions {
+              maybe_custom_root: None,
+            },
+          ))
+        },
+      ),
+      sys,
+      global_http_cache: Default::default(),
+      http_cache: Default::default(),
+      node_modules_dir_path: Default::default(),
+      npm_cache_dir: Default::default(),
+      npmrc: Default::default(),
+      node_modules_dir_mode: Default::default(),
+      workspace_directory: Default::default(),
+      initial_cwd,
+      options,
+    }
+  }
+
+  pub fn set_workspace_directory(
+    &mut self,
+    workspace_directory: WorkspaceDirectoryRc,
+  ) {
+    self.workspace_directory = Deferred::from(workspace_directory);
+  }
+
+  pub fn initial_cwd(&self) -> &PathBuf {
+    &self.initial_cwd
+  }
+
+  pub fn no_npm(&self) -> bool {
+    self.options.no_npm
+  }
+
+  pub fn node_modules_dir_mode(
```

Comment:
Pay attention here. I rewrote the logic for determining the NodeModulesDirMode because previously it was convoluted. This should be much more straightforward.